sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2164,16 +2164,16 @@ class Dataset[T] private[sql](
       columnMap.find { case (colName, _) =>
         resolver(field.name, colName)
       } match {
-        case Some((colName: String, col: Column)) => col.as(colName)
-        case _ => Column(field)
+        case Some((colName: String, col: Column)) => col.as(colName).named
+        case _ => field
       }
     }
 
-    val newColumns = columnMap.filter { case (colName, col) =>
+    val newColumns = columnMap.filter { case (colName, _) =>
       !output.exists(f => resolver(f.name, colName))
-    }.map { case (colName, col) => col.as(colName) }
+    }.map { case (colName, col) => col.as(colName).named }
 
-    select(replacedAndExistingColumns ++ newColumns : _*)
+    CollapseProject(Project(replacedAndExistingColumns ++ newColumns, logicalPlan))
```

Comment:
Can we reduce the scope of this optimization? e.g. if the root node of this query is `Project`, update its project list to include `withColumns`, otherwise add a new Project.

Reply:
I don't think we can do that. Imagine the case when all the columns depend on the previously added one: if we would do that, we would end up with an invalid plan. Or am I missing something?